mailbox/event/json/src/main/scala/org/apache/james/event/json/DTOs.scala

```diff
@@ -76,56 +76,71 @@ object DTOs {
     def fromJava(javaMessageMetaData: JavaMessageMetaData): MessageMetaData = DTOs.MessageMetaData(
       javaMessageMetaData.getUid,
       javaMessageMetaData.getModSeq,
-      javaMessageMetaData.getFlags,
+      Flags.fromJavaFlags(javaMessageMetaData.getFlags),
       javaMessageMetaData.getSize,
       javaMessageMetaData.getInternalDate.toInstant,
       javaMessageMetaData.getMessageId)
   }
 
-  case class MessageMetaData(uid: MessageUid, modSeq: Long, flags: JavaMailFlags, size: Long, internalDate: Instant, messageId: MessageId) {
-    def toJava: JavaMessageMetaData = new JavaMessageMetaData(uid, modSeq, flags, size, Date.from(internalDate), messageId)
+  case class MessageMetaData(uid: MessageUid, modSeq: Long, flags: Flags, size: Long, internalDate: Instant, messageId: MessageId) {
+    def toJava: JavaMessageMetaData = new JavaMessageMetaData(uid, modSeq, Flags.toJavaFlags(flags), size, Date.from(internalDate), messageId)
   }
 
   sealed trait Flag
 
   case class UserFlag(value: String) extends Flag
 
-  trait SystemFlag extends Flag
+  trait SystemFlag extends Flag {
+    def asString: String
```

Comment:
I dislike this asString thing. It's a DTO, we can very well use the type name IMO, it would make everything shorter, don't you think ?

Reply:
I had hard time getting the class name of your ADT:

getCanonicalName : 

```
java.lang.InternalError: Malformed class name
```

getSimpleName : 

```
java.lang.InternalError: Malformed class name
```

getTypeName:

```
org.apache.james.event.json.DTOs$SystemFlag$Deleted
```

Moving the ADT as TLD component of the class solved this 

however pattern matching on the class names is not supported: we should have statically defined strings...

Hence:

![capture d ecran de 2018-12-20 15-31-42](https://user-images.githubusercontent.com/6928740/50273238-b15aae80-046c-11e9-8649-6576fae52d0c.png)


Could we consider the following version to be working & mergeable and your proposal as a **non essential refactoring**???